sway-lsp/src/core/token.rs

```diff
@@ -162,22 +163,60 @@ impl Token {
         }
     }
 
-    /// Return the [Ident] of the declaration of the provided token.
-    pub fn declared_token_ident(&self, engines: &Engines) -> Option<Ident> {
+    /// Return the [TokenIdent] of the declaration of the provided token.
+    pub fn declared_token_ident(&self, engines: &Engines) -> Option<TokenIdent> {
         self.type_def.as_ref().and_then(|type_def| match type_def {
             TypeDefinition::TypeId(type_id) => ident_of_type_id(engines, type_id),
-            TypeDefinition::Ident(ident) => Some(ident.clone()),
+            TypeDefinition::Ident(ident) => Some(TokenIdent::new(ident, engines.se())),
         })
     }
+}
 
-    /// Return the [Span] of the declaration of the provided token. This is useful for
-    /// performaing == comparisons on spans. We need to do this instead of comparing
-    /// the [Ident] because the [PartialEq] implementation is only comparing the name.
-    pub fn declared_token_span(&self, engines: &Engines) -> Option<Span> {
-        self.type_def.as_ref().and_then(|type_def| match type_def {
-            TypeDefinition::TypeId(type_id) => Some(ident_of_type_id(engines, type_id)?.span()),
-            TypeDefinition::Ident(ident) => Some(ident.span()),
-        })
+/// A more convenient [Ident] type for use in the language server.
+///
+/// This type is used as the key in the [TokenMap]. It's constructed during AST traversal
+/// where we compute the [Range] of the token and the convert [SourceId]'s to [PathBuf]'s.
+/// Although this introduces a small amount of overhead while traversing, precomputing this
+/// greatly speeds up performace in all other areas of the language server.
+///
+/// [TokenMap]: crate::core::token_map::TokenMap
+/// [SourceId]: sway_types::SourceId
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub struct TokenIdent {
+    pub name: String,
+    pub range: Range,
+    pub path: Option<PathBuf>,
```

Comment:
Shouldn't an ident always have a path?

Reply:
Yeah I believe we should. I don't think it should be unreasonable to expect that all tokens in the AST's have a path. In another issue i'm running into errors where i have the following

```
std::module::MyType
```

The `std` token in this instance has no `SourceId`. This means I have no way of allowing go_to or hover on this token as the path needs to be known. I would be in favour of have a test in the compiler that makes sure all tokens that make it into the AST have a path. 

Keen to hear compiler team thoughts on this. cc @IGI-111 @tritao

Reply:
Have been thinking a bit more on this over the weekend. In some situations it actually does make sense for a node to exist in the AST that contains no span, and therefor no path. Namely, any token that is either a keyword or an intrinsic.

Keywords exist in the `LexedProgram` and intrinsics make it to the `ParseProgram` AST. So we already don't expect these tokens to have type definitions from a location. It feels like we should continue with having path  being an `Option<PathBuf>` for this reason. 

However, I would be in favour on still testing that all tokens that make it to the `TypedProgram` AST do in fact contain a valid path by checking if `span.source_id.is_some()`.